# Data Science Aptitude Test - Daimler Group - Part II

- date: yyyy.MM.dd
- author: [name](https://www.linkedin.com/in/)
***
## 1. Background 

The following test consists in one scenario in which you will have to analyze and train a model for one data set. The data set contains information about year, price, transmission, mileage, fuel type and engine size of used C-class cars. The idea is training a model to **predict** which will be the price of a used *C-class car* in the market. 

### Data Science Workflow stages

The competition solution workflow goes through ten stages described in the Data Science Solutions book.

1. Check Environment & libraries version.
2. Import Libraries.
3. Question or problem definition.
4. Acquire dataset or training and testing data.
5. Analyze, identify patterns, and explore the data.
6. Wrangle, prepare, cleanse the data.
7. Feature Engineering
8. **Model, predict and MLOps.**
9. **Visualize, report, and present the problem solving steps and final solution.**
10. Create or solve AI pipeline on Cloud, architecture solution and shows results.

Could be to exist another substeps during Data Processing and Data Modeling, for to train and deploy many machine learning and/or deep learning combination pipelines.

***
## 2. This notebook covers only:

1. [ ] Upload the csv file to your workspace and load it into a data frame.
2. [ ] Look for null values and outliers. Remove, keep or impute them and explain why you did so.
3. [ ] Show the main statistics (mean, standard deviation…) of the numerical columns of the data set. Are any of the variables skewed? (You can use any visualization you need to answer the last question).
4. [X] Train a model for the prediction of the price. Explain why you chose the model that you have trained.
5. [X] Test the model and obtain some performance metrics from it. Would you say that the model has a good performance? Why?
6. [X] Would you say that you have enough information to predict the price of an EQC (electric C-class)? Why?

## 3. Loading data from parque table

In [0]:
%sql


price transmission mileage fuelType engineSize carAge 30495 Automatic 1200 Diesel 2.0 1.0 29989 Automatic 1000 Petrol 1.5 1.0 37899 Automatic 500 Diesel 2.0 1.0

In [0]:
%sql


database tableName isTemporary default cclass_cleaned false

In [0]:
%sql


database tableName isTemporary information default cclass_cleaned false Database: default
Table: cclass_cleaned
Owner: root
Created Time: Tue Jul 20 10:50:59 UTC 2021
Last Access: UNKNOWN
Created By: Spark 3.1.1
Type: MANAGED
Provider: parquet
Statistics: 37055 bytes
Location: dbfs:/user/hive/warehouse/cclass_cleaned
Serde Library: org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe
InputFormat: org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat
OutputFormat: org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat
Schema: root
 |-- price: integer (nullable = true)
 |-- transmission: string (nullable = true)
 |-- mileage: integer (nullable = true)
 |-- fuelType: string (nullable = true)
 |-- engineSize: string (nullable = true)
 |-- carAge: double (nullable = true)

In [0]:
%scala
// Load dataframe from parquet table in Hive 


hivePath: String = dbfs:/user/hive/warehouse/cclass_cleaned
scalaDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 4 more fields]

In [0]:
%scala


price transmission mileage fuelType engineSize carAge 30495 Automatic 1200 Diesel 2.0 1.0 29989 Automatic 1000 Petrol 1.5 1.0 37899 Automatic 500 Diesel 2.0 1.0 30399 Automatic 5000 Diesel 2.0 2.0 29899 Automatic 4500 Diesel 2.0 2.0 35999 Automatic 500 Diesel 2.0 1.0 37990 Automatic 1412 Petrol 3.0 2.0 28990 Automatic 3569 Diesel 2.0 2.0 28990 Automatic 3635 Diesel 2.0 2.0 9995 Automatic 44900 Petrol 1.6 8.0 6995 Automatic 88200 Diesel 2.1 9.0 7495 Automatic 115000 Diesel 2.1 9.0 8995 Automatic 69250 Diesel 2.1 10.0 14995 Automatic 49850 Diesel 2.1 6.0 8595 Automatic 82685 Diesel 2.1 8.0 19250 Semi-Auto 32506 Diesel 2.1 5.0 23500 Semi-Auto 17000 Diesel 2.1 4.0 24995 Semi-Auto 17744 Diesel 2.0 3.0 29995 Semi-Auto 64291 Petrol 4.0 5.0 17990 Automatic 41011 Petrol 2.0 7.0 10977 Automatic 72000 Diesel 2.1 8.0 17977 Automatic 40000 Diesel 1.6 5.0 12790 Automatic 52000 Diesel 2.1 6.0 14488 Automatic 66472 Diesel 2.1 4.0 7500 Manual 91000 Diesel 2.1 10.0 9750 Automatic 69700 Diesel 2.1 8.0 15995 Automatic 39000 Diesel 1.6 6.0 15675 Manual 39287 Diesel 2.1 5.0 16475 Manual 40569 Diesel 2.1 5.0 15975 Manual 37419 Petrol 2.0 5.0 16475 Automatic 34452 Diesel 2.1 5.0 15975 Automatic 41567 Hybrid 2.1 6.0 17795 Automatic 37778 Diesel 2.1 4.0 17875 Automatic 44325 Diesel 2.1 4.0 14000 Automatic 45000 Diesel 2.1 5.0 16500 Automatic 75000 Diesel 1.6 4.0 21000 Automatic 36000 Diesel 2.1 5.0 15500 Automatic 64000 Diesel 2.1 6.0 7250 Manual 106000 Diesel 2.1 8.0 13000 Automatic 67150 Diesel 2.1 6.0 11000 Automatic 43426 Petrol 1.6 9.0 13300 Manual 42622 Diesel 2.1 5.0 16000 Manual 34903 Diesel 2.1 5.0 12500 Automatic 35345 Diesel 2.1 7.0 14495 Manual 23146 Petrol 2.0 6.0 24500 Automatic 16000 Diesel 2.1 4.0 23990 Automatic 44000 Petrol 6.2 8.0 17250 Automatic 30000 Diesel 2.1 6.0 14200 Automatic 22000 Petrol 1.6 6.0 22950 Automatic 19000 Petrol 2.0 4.0 10500 Automatic 90000 Diesel 2.1 8.0 15650 Automatic 32000 Diesel 2.1 7.0 14850 Automatic 90000 Diesel 2.1 6.0 22400 Automatic 37000 Diesel 2.1 4.0 19950 Automatic 23000 Diesel 2.1 4.0 9990 Automatic 78000 Diesel 2.1 7.0 11990 Automatic 47000 Diesel 2.1 6.0 20990 Automatic 18000 Diesel 2.1 4.0 14490 Automatic 64000 Diesel 2.1 6.0 10490 Automatic 50000 Diesel 2.1 8.0 15990 Automatic 65000 Diesel 2.1 5.0 17850 Automatic 38000 Diesel 2.1 5.0 16450 Automatic 70000 Diesel 2.1 5.0 11350 Automatic 69000 Diesel 2.1 8.0 12900 Automatic 41000 Petrol 1.6 7.0 26000 Automatic 16250 Petrol 1.5 3.0 22767 Automatic 30676 Diesel 2.1 4.0 12000 Automatic 40005 Diesel 2.1 6.0 18999 Automatic 38000 Hybrid 2.0 4.0 14000 Automatic 45000 Diesel 2.1 5.0 12000 Automatic 40005 Diesel 2.1 6.0 4343 Automatic 120000 Diesel 2.1 12.0 16500 Automatic 51936 Diesel 2.1 5.0 27750 Automatic 4000 Diesel 2.0 2.0 28950 Automatic 8500 Diesel 2.0 2.0 3495 Automatic 119168 Petrol 1.8 14.0 59740 Automatic 21800 Petrol 4.0 4.0 44990 Automatic 19999 Petrol 4.0 2.0 54990 Automatic 1600 Petrol 4.0 2.0 15995 Automatic 46221 Diesel 2.1 6.0 14000 Automatic 45000 Diesel 2.1 5.0 12000 Automatic 40005 Diesel 2.1 6.0 22767 Automatic 30676 Diesel 2.1 4.0 12998 Semi-Auto 52712 Diesel 2.1 7.0 17998 Semi-Auto 59858 Diesel 2.1 4.0 28998 Manual 22 Petrol 1.6 2.0 24998 Semi-Auto 28234 Petrol 2.0 4.0 19498 Semi-Auto 22803 Diesel 2.1 5.0 23498 Semi-Auto 17093 Diesel 2.1 3.0 25498 Semi-Auto 8150 Diesel 1.6 2.0 24998 Semi-Auto 9344 Diesel 2.0 2.0 15498 Semi-Auto 36243 Diesel 2.1 7.0 20995 Semi-Auto 33000 Diesel 2.1 4.0 31995 Semi-Auto 11300 Petrol 3.0 3.0 23995 Semi-Auto 32800 Diesel 2.1 4.0 16995 Automatic 37600 Diesel 2.1 7.0 24995 Automatic 14000 Petrol 2.0 4.0 26998 Semi-Auto 24175 Petrol 6.2 7.0 20498 Semi-Auto 27968 Diesel 2.1 5.0 27498 Semi-Auto 9255 Diesel 2.0 3.0 16998 Semi-Auto 26973 Diesel 1.6 5.0 27498 Semi-Auto 12491 Diesel 2.1 3.0 27698 Semi-Auto 14999 Diesel 2.0 2.0 27498 Semi-Auto 37941 Petrol 3.0 5.0 32498 Semi-Auto 11301 Diesel 2.0 2.0 24498 Semi-Auto 12629 Diesel 2.1 2.0 21498 Semi-Auto 20981 Diesel 2.1 4.0 18998 Semi-Auto 23760 Diesel 2.1 4.0 

***
## 4. Data Modeling

### Working With Categorical Data

Now that we've prepped everything that we're going to be running. We need to prep our transformation to transform the data into something that we can use in our model. There are a couple of changes that we're going to make.

Because we have categorical data in several of our columns, we're going to have to convert those into a numerical representation.

In [0]:
%scala
// importing scala libraries



import org.apache.spark.sql.functions.log
import org.apache.spark.ml.feature.{StringIndexer, VectorAssembler}
import org.apache.spark.ml.Pipeline

### Creating Training and Test Data Sets

Before we begin feature engineering and modeling, we will divide our data set into two groups: train and test. Depending on the size of your data set, your train/test ratio may vary, but many data scientists use 80/20 as a standard train/test split.

In [0]:
%scala


There are 3166 rows in the training set, and 
721 in the test set
trainDF: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [price: int, transmission: string ... 5 more fields]
testDF: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [price: int, transmission: string ... 5 more fields]

> Splitted dataset in train and test part with 80-20 proportion.

### Data Preparation

Let's go ahead and index all of our categorical features, and set our label to be `log(price)`.

In [0]:
%scala


categoricalCols: Array[String] = Array(transmission, fuelType, engineSize)
indexOutputCols: Array[String] = Array(transmissionIndex, fuelTypeIndex, engineSizeIndex)

### One-hot encoding

In the pipeline we just created, we only had two stages, and our linear regression model only used one feature. Let’s take a look at how to build a slightly more complex pipeline that incorporates all of our numeric and categorical features.

Most machine learning models in MLlib expect numerical values as input, represented as vectors. To convert categorical values into numeric values, we can use a technique called one-hot encoding (OHE).

There are a few ways to one-hot encode your data with Spark. A common approach is to use the StringIndexer and OneHotEncoder. With this approach, the first step is to apply the StringIndexer estimator to convert categorical values into category indices. These category indices are ordered by label frequencies, so the most frequent label gets index 0, which provides us with reproducible results across various runs of the same data.

In [0]:
%scala


stringIndexer: org.apache.spark.ml.feature.StringIndexer = strIdx_b080f837e395
numericCols: Array[String] = Array(carAge)
assemblerInputs: Array[String] = Array(transmissionIndex, fuelTypeIndex, engineSizeIndex, carAge)
vecAssembler: org.apache.spark.ml.feature.VectorAssembler = VectorAssembler: uid=vecAssembler_ac4394f5d0f0, handleInvalid=error, numInputCols=4

### Creating a Pipeline

If we want to apply our model to our test set, then we need to prepare that data in the same way as the training set (i.e., pass it through the vector assembler). Oftentimes data preparation pipelines will have multiple steps, and it becomes cumbersome to remember not only which steps to apply, but also the ordering of the steps.

In [0]:
%scala


pipeline: org.apache.spark.ml.Pipeline = pipeline_5f5858c5abc5

### Build a baseline model

This task seems well suited to a random forest classifier, since the output is binary and there may be interactions between multiple variables.

The following code builds a simple classifier using scikit-learn. It uses MLflow to keep track of the model accuracy, and to save the model for later use.

In [0]:
import mlflow
import mlflow.pyfunc
import mlflow.sklearn
from mlflow.models.signature import infer_signature
from mlflow.utils.environment import _mlflow_conda_env
import cloudpickle
import time

## 4. XGBoost

Distributed XGBoost with Spark only has a Scala API, so we are going to create views of our DataFrames to use in Scala, as well as save our (untrained) pipeline to load in to Scala.

XGBoost is a specific implementation of the Gradient Boosting method which uses more accurate approximations to find the best tree model. It employs a number of nifty tricks that make it exceptionally successful, particularly with structured data. XGBoost has additional advantages: training is very fast and can be parallelized / distributed across clusters. Therefore, XGBoost was another model that is used in this study.

In [0]:
%scala


fileName: String = /tmp/xgboost_feature_pipeline

### Training and Evaluation

Because we specified several different variations of the model, we're going to need to evaluate them somehow. In order to do this we're going to need to set a regression evaluator as well as perform a train test split so that we can train our models [ie each one with a different set of parameters] and evaluate them in the same way as well as in an automated fashion.

In [0]:
%scala
// Let's load in our data/pipeline that we defined in Python. 
import org.apache.spark.ml.Pipeline


import org.apache.spark.ml.Pipeline
fileName: String = tmp/xgboost_feature_pipeline
pipeline: org.apache.spark.ml.Pipeline = pipeline_e59fa3a20b24
trainDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 5 more fields]
testDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 5 more fields]

> Now we are ready to train our XGBoost model!

In [0]:
%scala

import ml.dmlc.xgboost4j.scala.spark._
import org.apache.spark.sql.functions._



Tracker started, with env={DMLC_NUM_SERVER=0, DMLC_TRACKER_URI=10.172.253.166, DMLC_TRACKER_PORT=9091, DMLC_NUM_WORKER=1}
import ml.dmlc.xgboost4j.scala.spark._
import org.apache.spark.sql.functions._
paramMap: scala.collection.immutable.Map[String,AnyVal] = Map(seed -> 42, num_round -> 100, missing -> 0, eta -> 0.1, max_leaf_nodes -> 50)
xgboostEstimator: ml.dmlc.xgboost4j.scala.spark.XGBoostRegressor = xgbr_87a4a373b534
xgboostPipeline: org.apache.spark.ml.Pipeline = pipeline_29d1f609c41e
xgboostPipelineModel: org.apache.spark.ml.PipelineModel = pipeline_29d1f609c41e
xgboostLogPredictedDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 10 more fields]
expXgboostDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 10 more fields]

### Evaluate

Now we can evaluate how well our XGBoost model performed.

In [0]:
%scala


## 5. Evaluate ML Model

The following statistics are shown for your model:

- Mean Absolute Error (MAE): The average of absolute errors. An error is the difference between the predicted value and the actual value.
- Root Mean Squared Error (RMSE): The square root of the average of squared errors of predictions made on the test dataset.
- Relative Absolute Error: The average of absolute errors relative to the absolute difference between actual values and the average of all actual values.
- Relative Squared Error: The average of squared errors relative to the squared difference between the actual values and the average of all actual values.
- Coefficient of Determination: Also known as the R squared value, this statistical metric indicates how well a model fits the data.

For each of the error statistics, smaller is better. A smaller value indicates that the predictions are closer to the actual values. For the coefficient of determination, the closer its value is to one (1.0), the better the predictions.

In [0]:
%scala


// Shows results and evaluation metrics


RMSE is 3129.343627416227
R2 is 0.8665786691566443
*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
import org.apache.spark.ml.evaluation.RegressionEvaluator
regressionEvaluator: org.apache.spark.ml.evaluation.RegressionEvaluator = RegressionEvaluator: uid=regEval_6ff4ed953790, metricName=r2, throughOrigin=false
rmse: Double = 3129.343627416227
r2: Double = 0.8665786691566443

> Results not bad, R2 is 0.87

In [0]:
%scala
// Shows first 20 predictions


+-------------------+-----+------------------+
 features|price| prediction|
+-------------------+-----+------------------+
[1.0,0.0,13.0,19.0]| 1495|1299.6947155788614|
 [1.0,0.0,0.0,18.0]| 1795|2130.7725307131823|
 [1.0,1.0,1.0,19.0]| 2140|1311.1550307961352|
 [1.0,0.0,0.0,13.0]| 3995| 4983.484102486172|
 [1.0,0.0,0.0,13.0]| 4295| 4983.484102486172|
 [1.0,1.0,6.0,13.0]| 5875| 4948.471550237936|
 [1.0,1.0,6.0,10.0]| 6995| 8902.416751446177|
 [1.0,0.0,0.0,11.0]| 6999|6528.5199774162265|
 [2.0,0.0,0.0,8.0]| 7000| 9377.688660343088|
 [2.0,0.0,0.0,8.0]| 7250| 9377.688660343088|
 [1.0,0.0,0.0,11.0]| 7490|6528.5199774162265|
 (4,[3],[8.0])| 7495|10510.859721560013|
 [1.0,1.0,6.0,10.0]| 7995| 8902.416751446177|
 [1.0,1.0,3.0,8.0]| 8250|11015.034294641477|
 [1.0,0.0,0.0,8.0]| 8595| 9867.167185457953|
 (4,[3],[8.0])| 9298|10510.859721560013|
 [2.0,0.0,0.0,6.0]| 9791| 12327.30348183262|
 [1.0,1.0,6.0,9.0]| 9850| 8878.236152795705|
 [1.0,1.0,3.0,8.0]| 9881|11015.034294641477|
 (4,[3],[7.0])| 9891|13691.707951462035|
+-------------------+-----+------------------+
only showing top 20 rows

### Export Model

In [0]:
%scala


nativeModelPath: String = xgboost_native_model
xgboostModel: ml.dmlc.xgboost4j.scala.spark.XGBoostRegressionModel = xgbr_87a4a373b534

### Question: What happens if you change your cluster configuration?

To test this out, try spinning up a cluster with just one worker, and another with two workers. NOTE: This data is quite small (one partition), and you will need to test it out with a larger dataset (e.g. 2+ partitions). However, in this code below, we will simply repartition our data to simulate how it could have been partitioned differently on a different cluster configuration, and see if we get the same number of data points in our training set.

## 6. Linear Regression

Now that we have prepared our data, we can use the `LinearRegression` estimator to build our first model [Python](https://spark.apache.org/docs/latest/api/python/pyspark.ml.html#pyspark.ml.regression.LinearRegression)/[Scala](https://spark.apache.org/docs/latest/api/scala/#org.apache.spark.ml.regression.LinearRegression). Estimators accept a DataFrame as input and return a model, and have a `.fit()` method.

### Vector Assembler

Linear Regression expects a column of Vector type as input.

We can easily get the values from the features column into a single vector using VectorAssembler Python/Scala. VectorAssembler is an example of a transformer. Transformers take in a DataFrame, and return a new DataFrame with one or more columns appended to it. They do not learn from your data, but apply rule based transformations.

In [0]:
%scala


fileName: String = /tmp/lr_feature_pipeline

In [0]:
%scala



import org.apache.spark.ml.regression.LinearRegression
lr: org.apache.spark.ml.regression.LinearRegression = linReg_d612b3364537
lrPipeline: org.apache.spark.ml.Pipeline = pipeline_7843516d3e18
lrPipelineModel: org.apache.spark.ml.PipelineModel = pipeline_7843516d3e18
lrLogPredictedDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 10 more fields]
lrDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 10 more fields]

### Apply to Test set

In [0]:
%scala


predDF: org.apache.spark.sql.DataFrame = [price: int, transmission: string ... 10 more fields]

In [0]:
%scala
// Shows first 20 predictions


+-------------------+-----+-------------------+
 features|price| prediction|
+-------------------+-----+-------------------+
[1.0,0.0,13.0,19.0]| 1495| 14213.613750454631|
 [1.0,0.0,0.0,18.0]| 1795|-16911.601126888625|
 [1.0,1.0,1.0,19.0]| 2140|-17627.937561476658|
 [1.0,0.0,0.0,13.0]| 3995| -3928.602633859391|
 [1.0,0.0,0.0,13.0]| 4295| -3928.602633859391|
 [1.0,1.0,6.0,13.0]| 5875| 10921.589313215776|
 [1.0,1.0,6.0,10.0]| 6995| 18711.388409033312|
 [1.0,0.0,0.0,11.0]| 6999| 1264.5967633522996|
 [2.0,0.0,0.0,8.0]| 7000| 8075.545659833744|
 [2.0,0.0,0.0,8.0]| 7250| 8075.545659833744|
 [1.0,0.0,0.0,11.0]| 7490| 1264.5967633522996|
 (4,[3],[8.0])| 7495| 10033.246058505938|
 [1.0,1.0,6.0,10.0]| 7995| 18711.388409033312|
 [1.0,1.0,3.0,8.0]| 8250| 16122.6305964106|
 [1.0,0.0,0.0,8.0]| 8595| 9054.395859169843|
 (4,[3],[8.0])| 9298| 10033.246058505938|
 [2.0,0.0,0.0,6.0]| 9791| 13268.745057045438|
 [1.0,1.0,6.0,9.0]| 9850| 21307.98810763916|
 [1.0,1.0,3.0,8.0]| 9881| 16122.6305964106|
 (4,[3],[7.0])| 9891| 12629.845757111783|
+-------------------+-----+-------------------+
only showing top 20 rows

### Inspect the model

In [0]:
%scala


// Shows results and evaluation metrics


RMSE is 5271.532469745133
R2 is 0.6213896078224412
*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
import org.apache.spark.ml.evaluation.RegressionEvaluator
regressionEvaluator: org.apache.spark.ml.evaluation.RegressionEvaluator = RegressionEvaluator: uid=regEval_c4297c880d26, metricName=r2, throughOrigin=false
rmse: Double = 5271.532469745133
r2: Double = 0.6213896078224412

> Linear regression R^2 result is 0.62, better than XGBoost. Despite the name R2 containing “squared,” R2 values range from negative infinity to 1. Let’s take a look at the math behind this metric. R2 is computed as follows:
R2=1−SSresSStot

where SStot is the total sum of squares if you always predict ȳ:
SStot=∑i=1n(yi−y¯)2

and SSres is the sum of residuals squared from your model predictions (also known as the sum of squared errors, which we used to compute the RMSE):
SSres=∑i=1n(yi−yˆi)2

If your model perfectly predicts every data point, then your SSres = 0, making your R2 = 1. And if your SSres = SStot, then the fraction is 1/1, so your R2 is 0. This is what happens if your model performs the same as always predicting the average value, ȳ.

But what if your model performs worse than always predicting ȳ and your SSres is really large? Then your R2 can actually be negative! If your R2 is negative, you should reevaluate your modeling process. The nice thing about using R2 is that you don’t necessarily need to define a baseline model to compare against.

> So how do we know if 5271.5 is a good value for the RMSE? 

There are various ways to interpret this value, one of which is to build a simple baseline model and compute its RMSE to compare against. A common baseline model for regression tasks is to compute the average value of the label on the training set ȳ (pronounced y-bar), then predict ȳ for every record in the test data set and compute the resulting RMSE

***
## Decision Tree

In the previous part, you were working with the parametric model, Linear Regression. We could do some more hyperparameter tuning with the linear regression model, but we're going to try tree based methods and see if our performance improves.

### Create Regressor

Now let's build a `DecisionTreeRegressor` with the default hyperparameters [Python](https://spark.apache.org/docs/latest/api/python/pyspark.ml.html#pyspark.ml.regression.DecisionTreeRegressor)/[Scala](https://spark.apache.org/docs/latest/api/scala/#org.apache.spark.ml.regression.DecisionTreeRegressor).

In [0]:
%scala


import org.apache.spark.ml.regression.DecisionTreeRegressor
dt: org.apache.spark.ml.regression.DecisionTreeRegressor = dtr_4a4b56511a34

### Fit Pipeline

In [0]:
%scala
import org.apache.spark.ml.Pipeline

// Combine stages into pipeline


// Uncomment to perform fit
// val pipelineModel = pipeline.fit(trainDF) 

import org.apache.spark.ml.Pipeline
stages: Array[org.apache.spark.ml.PipelineStage with org.apache.spark.ml.util.DefaultParamsWritable{def copy(extra: org.apache.spark.ml.param.ParamMap): org.apache.spark.ml.PipelineStage with org.apache.spark.ml.util.DefaultParamsWritable{def copy(extra: org.apache.spark.ml.param.ParamMap): org.apache.spark.ml.PipelineStage with org.apache.spark.ml.util.DefaultParamsWritable}}] = Array(strIdx_e9a7eee38139, VectorAssembler: uid=vecAssembler_2a0b7b7adce5, handleInvalid=error, numInputCols=4, dtr_4a4b56511a34)
pipeline: org.apache.spark.ml.Pipeline = pipeline_82a8e2faa50a

#### maxBins

What is this parameter [maxBins](https://spark.apache.org/docs/latest/api/python/pyspark.ml.html#pyspark.ml.regression.DecisionTreeRegressor.maxBins)? Let's take a look at the PLANET implementation of distributed decision trees (which Spark uses) and compare it to this paper called [Yggdrasil](https://cs.stanford.edu/~matei/papers/2016/nips_yggdrasil.pdf) by Matei Zaharia and others. This will help explain the `maxBins` parameter.

In Spark, data is partitioned by row. So when it needs to make a split, each worker has to compute summary statistics for every feature for  each split point. Then these summary statistics have to be aggregated (via tree reduce) for a split to be made.

Think about it: What if worker 1 had the value `32` but none of the others had it. How could you communicate how good of a split that would be? So, Spark has a maxBins parameter for discretizing continuous variables into buckets, but the number of buckets has to be as large as the number of categorical variables.

Let's go ahead and increase maxBins to `40`.

In [0]:
%scala


res30: dt.type = dtr_4a4b56511a34

In [0]:
%scala


import org.apache.spark.ml.Pipeline
stages: Array[org.apache.spark.ml.PipelineStage with org.apache.spark.ml.util.DefaultParamsWritable{def copy(extra: org.apache.spark.ml.param.ParamMap): org.apache.spark.ml.PipelineStage with org.apache.spark.ml.util.DefaultParamsWritable{def copy(extra: org.apache.spark.ml.param.ParamMap): org.apache.spark.ml.PipelineStage with org.apache.spark.ml.util.DefaultParamsWritable}}] = Array(strIdx_e9a7eee38139, VectorAssembler: uid=vecAssembler_2a0b7b7adce5, handleInvalid=error, numInputCols=4, dtr_4a4b56511a34)
pipeline: org.apache.spark.ml.Pipeline = pipeline_4861ec562e10
pipelineModel: org.apache.spark.ml.PipelineModel = pipeline_4861ec562e10

In [0]:
%scala


DecisionTreeRegressionModel: uid=dtr_4a4b56511a34, depth=5, numNodes=61, numFeatures=4
 If (feature 3 <= 3.5)
 If (feature 2 in {0.0,1.0,2.0,3.0})
 If (feature 2 in {0.0,2.0,3.0})
 If (feature 3 <= 2.5)
 If (feature 3 <= 1.5)
 Predict: 31980.117647058825
 Else (feature 3 > 1.5)
 Predict: 25870.388145315486
 Else (feature 3 > 2.5)
 If (feature 1 in {0.0})
 Predict: 21448.233333333334
 Else (feature 1 not in {0.0})
 Predict: 23024.261904761905
 Else (feature 2 not in {0.0,2.0,3.0})
 If (feature 3 <= 2.5)
 If (feature 3 <= 1.5)
 Predict: 34481.086956521736
 Else (feature 3 > 1.5)
 Predict: 29317.093023255813
 Else (feature 3 > 2.5)
 If (feature 1 in {1.0,2.0})
 Predict: 21715.955555555556
 Else (feature 1 not in {1.0,2.0})
 Predict: 26605.262295081968
 Else (feature 2 not in {0.0,1.0,2.0,3.0})
 If (feature 2 in {4.0})
 If (feature 3 <= 1.5)
 If (feature 0 in {1.0})
 Predict: 46956.666666666664
 Else (feature 0 not in {1.0})
 Predict: 47935.25
 Else (feature 3 > 1.5)
 If (feature 3 <= 2.5)
 Predict: 41762.53424657534
 Else (feature 3 > 2.5)
 Predict: 36391.333333333336
 Else (feature 2 not in {4.0})
 If (feature 3 <= 2.5)
 If (feature 0 in {1.0})
 Predict: 60355.529411764706
 Else (feature 0 not in {1.0})
 Predict: 66016.05263157895
 Else (feature 3 > 2.5)
 If (feature 0 in {0.0})
 Predict: 45421.666666666664
 Else (feature 0 not in {0.0})
 Predict: 46948.25
 Else (feature 3 > 3.5)
 If (feature 2 in {0.0,1.0,3.0,6.0,7.0,8.0,9.0,10.0,11.0,12.0,13.0,14.0})
 If (feature 3 <= 5.5)
 If (feature 0 in {2.0})
 If (feature 3 <= 4.5)
 Predict: 15608.181818181818
 Else (feature 3 > 4.5)
 Predict: 14576.666666666666
 Else (feature 0 not in {2.0})
 If (feature 3 <= 4.5)
 Predict: 20140.960084033613
 Else (feature 3 > 4.5)
 Predict: 18029.96335078534
 Else (feature 3 > 5.5)
 If (feature 3 <= 8.5)
 If (feature 2 in {0.0,1.0,3.0,6.0,8.0,10.0})
 Predict: 13977.80195599022
 Else (feature 2 not in {0.0,1.0,3.0,6.0,8.0,10.0})
 Predict: 23663.75
 Else (feature 3 > 8.5)
 If (feature 3 <= 10.5)
 Predict: 8642.542857142857
 Else (feature 3 > 10.5)
 Predict: 5144.948717948718
 Else (feature 2 not in {0.0,1.0,3.0,6.0,7.0,8.0,9.0,10.0,11.0,12.0,13.0,14.0})
 If (feature 3 <= 6.5)
 If (feature 2 in {4.0})
 If (feature 3 <= 4.5)
 Predict: 32153.2
 Else (feature 3 > 4.5)
 Predict: 30854.666666666668
 Else (feature 2 not in {4.0})
 If (feature 3 <= 4.5)
 Predict: 42746.10526315789
 Else (feature 3 > 4.5)
 Predict: 34974.846153846156
 Else (feature 3 > 6.5)
 If (feature 0 in {1.0})
 If (feature 1 in {1.0})
 Predict: 4450.0
 Else (feature 1 not in {1.0})
 Predict: 4480.0
 Else (feature 0 not in {1.0})
 Predict: 11995.0

dtModel: org.apache.spark.ml.regression.DecisionTreeRegressionModel = DecisionTreeRegressionModel: uid=dtr_4a4b56511a34, depth=5, numNodes=61, numFeatures=4

### Feature Importance

Let's go ahead and get the fitted decision tree model, and look at the feature importance scores.

In [0]:
%scala


dtModel: org.apache.spark.ml.regression.DecisionTreeRegressionModel = DecisionTreeRegressionModel: uid=dtr_4a4b56511a34, depth=5, numNodes=61, numFeatures=4
res32: org.apache.spark.ml.linalg.Vector = (4,[0,1,2,3],[0.0076267640658624685,0.003136232295786416,0.41490996185935414,0.5743270417789971])

### Interpreting Feature Importance

Hmmm... it's a little hard to know what feature 4 vs 11 is. Given that the feature importance scores are "small data", let's use Pandas to help us recover the original column names.

In [0]:
%scala


+-----------------+--------------------+
 feature| Importance|
+-----------------+--------------------+
 carAge| 0.5743270417789971|
 engineSizeIndex| 0.41490996185935414|
transmissionIndex|0.007626764065862...|
 fuelTypeIndex|0.003136232295786416|
+-----------------+--------------------+

dtModel: org.apache.spark.ml.regression.DecisionTreeRegressionModel = DecisionTreeRegressionModel: uid=dtr_4a4b56511a34, depth=5, numNodes=61, numFeatures=4
featureImp: Array[(String, Double)] = Array((transmissionIndex,0.0076267640658624685), (fuelTypeIndex,0.003136232295786416), (engineSizeIndex,0.41490996185935414), (carAge,0.5743270417789971))
columns: Array[String] = Array(feature, Importance)
featureImpDF: org.apache.spark.sql.DataFrame = [feature: string, Importance: double]

> Yes, carAge is most important feature in that machine learning problem.

### Apply model to test set

In [0]:
%scala


features price prediction Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 5.0, 2.0)) 67495 66016.05263157895 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 2.0)) 67000 60355.529411764706 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 2.0)) 66990 60355.529411764706 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 2.0)) 61495 60355.529411764706 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 5.0, 2.0)) 57998 66016.05263157895 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 2.0)) 53990 60355.529411764706 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 4.0, 1.0)) 49998 46956.666666666664 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 49499 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 49499 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 2.0)) 49280 60355.529411764706 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 48990 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 4.0, 2.0)) 46390 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 45980 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 45980 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 4.0, 2.0)) 45900 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 5.0, 3.0)) 45895 45421.666666666664 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 4.0)) 44520 42746.10526315789 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 44399 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 43199 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 42999 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 42995 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 4.0, 2.0)) 42895 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 42699 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 4.0, 2.0)) 42490 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 41699 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 41400 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 5.0, 4.0)) 41399 42746.10526315789 Map(vectorType -> dense, length -> 4, values -> List(0.0, 0.0, 1.0, 1.0)) 40991 34481.086956521736 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 40849 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 39995 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 0.0, 1.0, 1.0)) 39991 34481.086956521736 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 6.0)) 39990 34974.846153846156 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 5.0, 4.0)) 39336 42746.10526315789 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 4.0, 2.0)) 39299 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 2.0)) 38995 41762.53424657534 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 4.0, 3.0)) 38495 36391.333333333336 Map(vectorType -> dense, length -> 4, values -> List(0.0, 1.0, 4.0, 3.0)) 37999 36391.333333333336 Map(vectorType -> dense, length -> 4, values -> List(1.0, 1.0, 2.0, 2.0)) 37999 25870.388145315486 Map(vectorType -> dense, length -> 4, values -> List(1.0, 0.0, 1.0, 2.0)) 37990 29317.093023255813 Map(vectorType -> dense, length -> 4, values -> List(1.0, 0.0, 1.0, 2.0)) 37250 29317.093023255813 Map(vectorType -> dense, length -> 4, v

In [0]:
%scala


RMSE is 3325.055165515528
R2 is 0.8493682683711843
*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
import org.apache.spark.ml.evaluation.RegressionEvaluator
regressionEvaluator: org.apache.spark.ml.evaluation.RegressionEvaluator = RegressionEvaluator: uid=regEval_3ae24db6b521, metricName=r2, throughOrigin=false
rmse: Double = 3325.055165515528
r2: Double = 0.8493682683711843

> What will a decision tree predict?

It turns out decision trees cannot predict any values larger than they were trained on. The max value in our training set was €66,000, so we can't predict any values larger than that (or technically any values larger than the ).

- **R^2** is 0.85 like XGBoost (yes, that works same parameter and decision tree based.)

About **RMSE** first model and last model minimize the error between 3000 and 3300, instead LR model more 5000. We cannot used normalize or scaled dataset, in the next steps, we need adopted PCA, Scaling and Normalize values and compare the results again.

## 7. Summary

This model is worse than the linear regression model.
Next workstages will be check other features, for instance look at hyperparameter tuning and ensemble models to improve upon the performance of our singular decision tree.

There are some papers, used different models in order to predict used car prices. However, there was a relatively small dataset for making a strong inference because number of observations was only 380962. Gathering more data can yield more robust predictions. Secondly, there could be more features that can be good predictors. For example, here are some variables that might improve the model: number of doors, gas/mile (per gallon), color, mechanical and cosmetic reconditioning time, used-to-new ratio, appraisal-to-trade ratio.

Another point that that has room to improvement is that data cleaning process can be dome more rigorously with the help of more technical information. For example, instead of using ‘ffill’ method, there might be indicators that helps to fill missing values more meaningfully.

As suggestion for further studies, while pre-processing data, instead of using label encoder, one hot encoder method can be used. Thus, all non-numeric features can be converted to nominal data instead of ordinal data (Raschka & Mirjalili, 2017). This may cause a serious change in performance of predictive models. Also, after training the data, instead of min-max scaler, standard scaler can be performned and results can be compared. Different scalers can be checked whether there is an improvement in prediction power of models or not.

## 8. Would you say that you have enough information to predict the price of an EQC (electric C-class)? Why?

> The answer is not in part, only with few features. After scraped **autoscout.de**, in attachment we will found the notebook, there are more variables about same model C-Klasse and, in other for EQC cars. There are specific features apart mileage and age, most important features to determinate the car used price, like other important amenities and specific features for electric-car, yes are different and after a good analysis and understand the correlation, maybe we be able to create a similar machine learning model, adjust hyperparameter or create an ensemble model.